proxy/proxy.go

```diff
@@ -467,24 +470,60 @@ func setRequestURLForDynamicBackend(u *url.URL, stateBag map[string]interface{})
 	}
 }
 
-func selectEndpoint(ctx *context, registry *routing.EndpointRegistry) *routing.LBEndpoint {
+func fadeIn(now time.Time, duration time.Duration, exponent float64, detected time.Time) float64 {
+	rel := now.Sub(detected)
+	fadingIn := rel > 0 && rel < duration
+	if !fadingIn {
+		return 1
+	}
+
+	return math.Pow(float64(rel)/float64(duration), exponent)
+}
+
+func filterFadeIn(endpoints []routing.LBEndpoint, rt *routing.Route, registry *routing.EndpointRegistry, rnd *rand.Rand) []routing.LBEndpoint {
+	now := time.Now()
+	threshold := rnd.Float64()
+
+	filtered := make([]routing.LBEndpoint, 0, len(endpoints))
+	for _, e := range endpoints {
+		detected := registry.GetMetrics(e.Host).DetectedTime()
+		f := fadeIn(
+			now,
```

Comment:
we could also pass "rel" and drop "now" and "detected" from the input:
```
fadeIn( 
  now.Sub(registry.GetMetrics(e.Host).DetectedTime()),
  ...
)
```
this would omit some copies. Not sure if this would be inlined anyways, but maybe it's ok.

Reply:
I will provide some benchmarks to check whether it makes sense.

Reply:
There is no significant performance/memory difference between code provided before and after this comment.
However, code after comment has less amount of argument and therefore is more easy to read.

```
$ benchstat decoupleFadeIn.txt decoupleFadeInRel.txt 
goos: linux
goarch: amd64
pkg: github.com/zalando/skipper/proxy
cpu: AMD Ryzen 7 PRO 4750U with Radeon Graphics
                                                     │ decoupleFadeIn.txt │        decoupleFadeInRel.txt        │
                                                     │       sec/op       │   sec/op     vs base                │
FadeIn/power-of-n-random-choices,_11,_1_clients-16           35.11µ ±  3%   33.84µ ± 4%   -3.63% (p=0.000 n=60)
FadeIn/power-of-n-random-choices,_11,_4_clients-16           12.30µ ±  7%   11.24µ ± 6%   -8.60% (p=0.014 n=60)
FadeIn/power-of-n-random-choices,_11,_16_clients-16          4.934µ ±  1%   4.878µ ± 1%        ~ (p=0.115 n=60)
FadeIn/power-of-n-random-choices,_11,_64_clients-16          4.623µ ±  1%   4.504µ ± 1%   -2.57% (p=0.000 n=60)
FadeIn/power-of-n-random-choices,_11,_256_clients-16         4.174µ ±  1%   4.137µ ± 1%        ~ (p=0.208 n=60)
FadeIn/random,_11,_1_clients-16                              36.43µ ±  6%   35.45µ ± 3%        ~ (p=0.085 n=60)
FadeIn/random,_11,_4_clients-16                              12.58µ ± 13%   12.14µ ± 6%        ~ (p=0.865 n=60)
FadeIn/random,_11,_16_clients-16                             4.888µ ±  1%   4.835µ ± 2%        ~ (p=0.069 n=60)
FadeIn/random,_11,_64_clients-16                             4.674µ ±  1%   4.598µ ± 1%   -1.63% (p=0.000 n=60)
FadeIn/random,_11,_256_clients-16                            4.194µ ±  1%   4.150µ ± 1%        ~ (p=0.052 n=60)
FadeIn/round-robin,_11,_1_clients-16                         35.88µ ±  3%   34.84µ ± 3%   -2.90% (p=0.011 n=60)
FadeIn/round-robin,_11,_4_clients-16                         11.10µ ± 15%   12.31µ ± 5%        ~ (p=0.185 n=60)
FadeIn/round-robin,_11,_16_clients-16                        4.923µ ±  2%   4.854µ ± 1%        ~ (p=0.055 n=60)
FadeIn/round-robin,_11,_64_clients-16                        4.655µ ±  1%   4.608µ ± 1%   -1.01% (p=0.032 n=60)
FadeIn/round-robin,_11,_256_clients-16                       4.188µ ±  2%   4.057µ ± 1%   -3.13% (p=0.000 n=60)
FadeIn/consistent-hash,_11,_1_clients-16                     36.23µ ±  3%   36.73µ ± 2%        ~ (p=0.661 n=60)
FadeIn/consistent-hash,_11,_4_clients-16                     12.94µ ±  6%   11.51µ ± 4%  -11.07% (p=0.000 n=60)
FadeIn/consistent-hash,_11,_16_clients-16                    4.585µ ±  3%   4.521µ ± 1%        ~ (p=0.161 n=60)
FadeIn/consistent-hash,_11,_64_clients-16                    4.477µ ±  1%   4.354µ ± 1%   -2.75% (p=0.000 n=60)
FadeIn/consistent-hash,_11,_256_clients-16                   4.107µ ±  1%   4.001µ ± 1%   -2.59% (p=0.000 n=60)
geomean                                                      8.352µ         8.175µ        -2.11%

                                                     │ decoupleFadeIn.txt │        decoupleFadeInRel.txt        │
                                                     │        B/op        │     B/op      vs base               │
FadeIn/power-of-n-random-choices,_11,_1_clients-16           10.21Ki ± 0%   10.21Ki ± 0%   0.00% (p=0.012 n=60)
FadeIn/power-of-n-random-choices,_11,_4_clients-16           10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.572 n=60)
FadeIn/power-of-n-random-choices,_11,_16_clients-16          10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.483 n=60)
FadeIn/power-of-n-random-choices,_11,_64_clients-16          10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.316 n=60)
FadeIn/power-of-n-random-choices,_11,_256_clients-16         10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.067 n=60)
FadeIn/random,_11,_1_clients-16                              10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.325 n=60)
FadeIn/random,_11,_4_clients-16                              10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.147 n=60)
FadeIn/random,_11,_16_clients-16                             10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.718 n=60)
FadeIn/random,_11,_64_clients-16                             10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.791 n=60)
FadeIn/random,_11,_256_clients-16                            10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.900 n=60)
FadeIn/round-robin,_11,_1_clients-16                         10.21Ki ± 0%   10.21Ki ± 0%   0.00% (p=0.044 n=60)
FadeIn/round-robin,_11,_4_clients-16                         10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.858 n=60)
FadeIn/round-robin,_11,_16_clients-16                        10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.935 n=60)
FadeIn/round-robin,_11,_64_clients-16                        10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.212 n=60)
FadeIn/round-robin,_11,_256_clients-16                       10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.193 n=60)
FadeIn/consistent-hash,_11,_1_clients-16                     10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.325 n=60)
FadeIn/consistent-hash,_11,_4_clients-16                     10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.858 n=60)
FadeIn/consistent-hash,_11,_16_clients-16                    10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.173 n=60)
FadeIn/consistent-hash,_11,_64_clients-16                    10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.342 n=60)
FadeIn/consistent-hash,_11,_256_clients-16                   10.21Ki ± 0%   10.21Ki ± 0%       ~ (p=0.807 n=60)
geomean                                                      10.21Ki        10.21Ki       +0.00%

                                                     │ decoupleFadeIn.txt │       decoupleFadeInRel.txt        │
                                                     │     allocs/op      │  allocs/op   vs base               │
FadeIn/power-of-n-random-choices,_11,_1_clients-16            8.000 ± 12%   7.500 ±  7%       ~ (p=0.363 n=60)
FadeIn/power-of-n-random-choices,_11,_4_clients-16            8.000 ±  0%   8.000 ±  0%       ~ (p=0.655 n=60)
FadeIn/power-of-n-random-choices,_11,_16_clients-16           8.000 ±  0%   8.000 ±  0%       ~ (p=0.655 n=60)
FadeIn/power-of-n-random-choices,_11,_64_clients-16           8.000 ±  0%   8.000 ±  0%       ~ (p=0.403 n=60)
FadeIn/power-of-n-random-choices,_11,_256_clients-16          8.000 ± 12%   7.000 ± 14%       ~ (p=0.147 n=60)
FadeIn/random,_11,_1_clients-16                               7.000 ±  0%   7.000 ±  0%       ~ (p=1.000 n=60)
FadeIn/random,_11,_4_clients-16                               8.000 ±  0%   8.000 ±  0%       ~ (p=0.147 n=60)
FadeIn/random,_11,_16_clients-16                              8.000 ±  0%   8.000 ±  0%       ~ (p=0.315 n=60)
FadeIn/random,_11,_64_clients-16                              8.000 ±  0%   8.000 ±  0%       ~ (p=0.436 n=60)
FadeIn/random,_11,_256_clients-16                             8.000 ± 12%   8.000 ± 12%       ~ (p=0.463 n=60)
FadeIn/round-robin,_11,_1_clients-16                          7.000 ±  0%   7.000 ±  0%       ~ (p=0.159 n=60)
FadeIn/round-robin,_11,_4_clients-16                          8.000 ±  0%   8.000 ±  0%       ~ (p=0.757 n=60)
FadeIn/round-robin,_11,_16_clients-16                         8.000 ±  0%   8.000 ±  0%       ~ (p=1.000 n=60)
FadeIn/round-robin,_11,_64_clients-16                         8.000 ±  0%   8.000 ±  0%       ~ (p=0.147 n=60)
FadeIn/round-robin,_11,_256_clients-16                        7.500 ±  7%   8.000 ± 12%       ~ (p=0.363 n=60)
FadeIn/consistent-hash,_11,_1_clients-16                      7.000 ±  0%   7.000 ±  0%       ~ (p=0.325 n=60)
FadeIn/consistent-hash,_11,_4_clients-16                      8.000 ±  0%   8.000 ±  0%       ~ (p=0.437 n=60)
FadeIn/consistent-hash,_11,_16_clients-16                     8.000 ±  0%   8.000 ±  0%       ~ (p=0.227 n=60)
FadeIn/consistent-hash,_11,_64_clients-16                     8.000 ± 12%   8.000 ±  0%       ~ (p=0.186 n=60)
FadeIn/consistent-hash,_11,_256_clients-16                    7.000 ± 14%   8.000 ± 12%       ~ (p=0.587 n=60)
geomean                                                       7.764         7.764        +0.00%
```